### DISCUSSION TITLE
NumCPUs for the tor process

### DISCUSSION THREAD

**Discussion by Superpaul209** [Enkidu-6/tor-ddos#59] [Q&A]
Hello everyone,

I would like to know what is the best option to set for NumCPUs to optimize tor on my servers.
Im having an old (slow but enough to make a guard relay) desktop PC running with 4 cores, 4 GB RAM. How many NumCPUs should I put for this server ?
When looking nyx, its always between 100 and 125% of CPU usage which isn't good for overload status.

I also have many small servers running with only one core and 1 GB RAM for bridges, how many NumCPUs should I put for them ?

Thanks for your help 😄

**Comment by Enkidu-6** [comment]
Hi,
As I've mentioned before in one of our discussions here, NumCPUs is poorly documented. It has nothing to do with your actual number of CPUs. It's about the number of worker threads. Tor Worker threads are responsible for decrypting onionskins and Tor decides to keep or drop them based on how long it takes for a worker to become available. By increasing the NumCPUs you're creating available worker threads to be used when needed and until they're needed, they use zero CPU.

To deal with current attacks I recommend a minimum of 16 and that depends on the traffic you're relaying. I have mine around 24-30. Even though your system has 4 CPUs you can safely and easily set it to 16 or more. It's not like overclocking and it won't put much pressure on your CPU.

**Comment by Enkidu-6** [comment]
By the way, I suggest you start from a certain number and look at how you're dealing with the traffic and whether you show up as overloaded. Then gradually increase the number until you find that magic number.

I use htop to look at the worker threads during high loads and when I see a couple of them using 0% I know I have enough.

By the way, each time you change the number you have to restart Tor for it to take effect. A -HUP signal won't do.

**Comment by Superpaul209** [comment]
Thanks, I have put 16 processes to see how it goes. Its currently using 2% of CPU for each process (with htop) and main process uses 112%
I may need to put more processes so it will handles more resources. My guard relay has a maximum of 8 MB/s of traffic with average of 6 MB/s.

**Comment by Enkidu-6** [comment]
Yes. Tor doesn't understand multiple CPUs so it combines everything and reports it under one CPU. The 112% is actually the combination of all those worker threads plus the main process.  So it's actually using 1.12 CPUs out of 4 (or 8 if your processor is multi threaded)
